doc/benchmarking.rst

```diff
@@ -0,0 +1,160 @@
+Benchmarking
+============
+Performing a benchmark of fgpu or xbgpu under realistic conditions (in
+particular, with data transmitted over UDP) is difficult because there is no
+flow control, and the achievable rate can only be observed indirectly by
+running the system at a particular rate and checking whether it keeps up. We
+wish to know the largest rate at which we keep up most of the time; we'll call
+this the "critical" rate. The :file:`scratch/benchmarks` directory contains a
+script to help estimate the critical rate for fgpu (xbgpu support may be added
+later).
+
+To run it, you'll need
+
+- two servers (one to generate data and one to receive it)
+
+  - configured for use with :external+spead2:doc:`spead2's ibverbs support
+    <py-ibverbs>` on a high-speed network between them;
+  - with a Docker daemon;
+
+- somewhere to run the script (referred to below as the "client"), set up for
+  key-based SSH access to the servers in accounts that can invoke the Docker
+  client.
+- a Docker registry containing the katgpucbf Docker image.
+
+The assignment of threads to processor cores on the servers has been
+optimised for AMD Epyc servers with at least 16 cores; for other servers or
+for lower core counts, the benchmark script may need to be tuned to provide
+sensible placement.
+
+The client machine will also need a few Python packages installed, including
+:mod:`aiohttp`, :mod:`asyncssh`, :mod:`numpy` and :mod:`prometheus_client`
+(it's easiest to just run the script then install whatever packages are
+missing).
+
+On the client machine, you will need to create a TOML_ file describing the
+servers you want to use. Each table in the TOML file describes one server. You
+can give them any names you like, but by default the script will use the
+servers called ``dsim`` and ``fgpu``. A typical file looks like this:
+
+.. code-block:: toml
+
+   [dsim]
+   hostname = 'server01.domain'
+   username = 'myusername'
+   interfaces = ['enp193s0f0np0', 'enp193s0f1np1']
+
+   [fgpu]
+   hostname = 'server02.domain'
+   username = 'myusername'
+   interfaces = ['enp193s0f0np0', 'enp193s0f1np1']
+
+.. _TOML: https://toml.io/
+
+The `interfaces` tables list the names of the ibverbs-capable network
```

Comment:
nit-pick: I think this would be a key/value pair, not a table.

Reply:
I was referring to the value (`['enp193s0f0np0', 'enp193s0f1np1']`), but actually that isn't a table - it's an array. Will fix.